Sources/MapboxMaps/Viewport/Viewport.swift

```diff
@@ -0,0 +1,114 @@
+                  cameraAnimationsManager: CameraAnimationsManagerProtocol,
+                  mapboxMap: MapboxMapProtocol) {
+        self.impl = impl
+        self.locationProducer = locationProducer
+        self.cameraAnimationsManager = cameraAnimationsManager
+        self.mapboxMap = mapboxMap
+    }
+
+    // the list of states. order is not guaranteed
+    // TODO: should we guarantee order? Converting this to a Set would be inconvenient
+    public var states: [ViewportState] {
+        impl.states
+    }
+
+    // adds state to the list of states
+    // adding the same state more than once has no effect
+    public func addState(_ state: ViewportState) {
+        impl.addState(state)
+    }
+
+    // removes state from list of states
+    // sets current state to nil if it was identical to state
+    // attempting to remove a state that was not added has no effect
+    // any active transition to that state will also be canceled
+    public func removeState(_ state: ViewportState) {
+        impl.removeState(state)
+    }
+
+    // MARK: - Current State
+
+    // a nil status is known as "idle"; this is the default
+    public var status: ViewportStatus? {
+        impl.status
+    }
+
+    // set
+    // the Bool in the completion block indicates whether the transition ran to
+    // completion (true) or was interrupted by another transition (false)
+    public func transition(to toState: ViewportState?, completion: ((Bool) -> Void)? = nil) {
+        impl.transition(to: toState, completion: completion)
+    }
+
+    // MARK: - Transitions
+
+    // this transition is used unless overridden by one of the registered transitions
+    public var defaultTransition: ViewportTransition {
+        get { impl.defaultTransition }
+        set { impl.defaultTransition = newValue }
+    }
+
+    // set
+    // we allow setting a custom transition from idle (nil) to a state, but
+    // there's never a transition when going from some non-nil state to idle.
+    public func setTransition(_ transition: ViewportTransition,
+                              from fromState: ViewportState?,
+                              to toState: ViewportState) {
+        impl.setTransition(transition, from: fromState, to: toState)
+    }
+
+    // get
+    public func getTransition(from fromState: ViewportState?,
+                              to toState: ViewportState) -> ViewportTransition? {
+        impl.getTransition(from: fromState, to: toState)
+    }
+
+    // delete
+    public func removeTransition(from fromState: ViewportState?,
+                                 to toState: ViewportState) {
+        impl.removeTransition(from: fromState, to: toState)
+    }
+
+    // factory methods
+
+    public func makeFollowingViewportState(options: FollowingViewportStateOptions = .init()) -> FollowingViewportState {
+        return FollowingViewportState(
+            options: options,
+            locationProducer: locationProducer,
```

Comment:
Would it make sense to allow users to provide their own generic location providers? At this point Navigation SDK API listens for changes in multiple sources (raw locations from Maps SDK and enhanced locations from Navigation Native).

I think it would be also beneficial to switch sources on the fly, without creating state object.

Reply:
I've added this to a list of topics to revisit in between the beta and RC releases.